File: oc/Quartz2DDemo/Quartz2DDemo/HuangView.m
//
//  HuangView.m
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-1.
//  Copyright © 2018年 east. All rights reserved.
//

#import "HuangView.h"

@implementation HuangView

- (void)drawRect:(CGRect)rect {
    CGFloat marginTop = 50;
    CGFloat radius = 80;
    CGFloat middleH = 150;
    CGFloat glassH = 30;
    CGFloat rectW = rect.size.width;
    
    // 获取上下文
    CGContextRef context = UIGraphicsGetCurrentContext();
    // 设置填充颜色为黄色
//    [[UIColor yellowColor] set];
    CGContextSetRGBFillColor(context, 1.0, 1.0, 0, 1);
    // 绘制小黄人头部半圆
    CGContextAddArc(context, rectW * 0.5, marginTop + radius, radius, 0, M_PI, 1);
    // 绘制小黄人身体
    CGContextAddRect(context, CGRectMake(rectW * 0.5 - radius, marginTop + radius, radius * 2, middleH));
    // 绘制小黄人底部
    CGContextAddArc(context, rectW * 0.5, marginTop + radius + middleH, radius, 0, M_PI, 0);
    // 渲染
    CGContextFillPath(context);
    
    // 更改填充颜色为黑色
    CGContextSetRGBFillColor(context, 0, 0, 0, 1);
    // 绘制眼眶
    CGFloat padding = 5;
    CGContextAddRect(context, CGRectMake(rectW * 0.5 - radius - padding, marginTop + radius, (radius + padding) * 2, glassH));
    CGContextFillPath(context);
    
    // 更改填充颜色为白色
    CGContextSetRGBFillColor(context, 1.0, 1.0, 1.0, 1);
    CGFloat whiteRadius = radius * 0.4;
    CGContextAddArc(context, rectW * 0.5 - whiteRadius, marginTop + radius + glassH * 0.5, whiteRadius, 0, 2 * M_PI, 0);
    CGContextAddArc(context, rectW * 0.5 + whiteRadius, marginTop + radius + glassH * 0.5, whiteRadius, 0, 2 * M_PI, 0);
    CGContextFillPath(context);
    
    // 设置线条宽度为10
    CGContextSetLineWidth(context, 10);
    // 更改线条颜色为黑线
    CGContextSetRGBStrokeColor(context, 0, 0, 0, 1);
    CGContextAddArc(context, rectW * 0.5 - whiteRadius, marginTop + radius + glassH * 0.5, whiteRadius, 0, 2 * M_PI, 0);
    CGContextMoveToPoint(context,  rectW * 0.5 + whiteRadius * 2, marginTop + radius + glassH * 0.5);
    CGContextAddArc(context, rectW * 0.5 + whiteRadius, marginTop + radius + glassH * 0.5, whiteRadius, 0, 2 * M_PI, 0);
    // 渲染
    CGContextStrokePath(context);
    
    // 绘制眼球
    CGContextSetRGBFillColor(context, 139/255.0, 105/255.0, 105/255.0, 1);
    CGFloat eyeballRadius = radius * 0.2;
    CGContextAddArc(context, rectW * 0.5 - eyeballRadius - 10, marginTop + radius + glassH * 0.5, eyeballRadius, 0, 2 * M_PI, 0);
    CGContextAddArc(context, rectW * 0.5 + eyeballRadius + 10, marginTop + radius + glassH * 0.5, eyeballRadius, 0, 2 * M_PI, 0);
    CGContextFillPath(context);
    
    CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1);
    CGContextAddArc(context, rectW * 0.5 - eyeballRadius - 10, marginTop + radius + glassH * 0.5, eyeballRadius * 0.5, 0, 2 * M_PI, 0);
    CGContextAddArc(context, rectW * 0.5 + eyeballRadius + 10, marginTop + radius + glassH * 0.5, eyeballRadius * 0.5, 0, 2 * M_PI, 0);
    CGContextFillPath(context);
    
    CGContextSetRGBFillColor(context, 1.0, 1.0, 1.0, 1);
    CGContextAddArc(context, rectW * 0.5 - eyeballRadius * 1.1 - 10, marginTop + radius + glassH * 0.5 - eyeballRadius * 0.1, eyeballRadius * 0.1, 0, 2 * M_PI, 0);
    CGContextAddArc(context, rectW * 0.5 + eyeballRadius * 0.9 + 10, marginTop + radius + glassH * 0.5 - eyeballRadius * 0.1, eyeballRadius * 0.1, 0, 2 * M_PI, 0);
    CGContextFillPath(context);
    
    // 嘴巴
    CGContextSetRGBStrokeColor(context, 0, 0, 0, 1);
    CGContextSetLineWidth(context, 2);
    CGContextAddArc(context, rectW * 0.5, marginTop + radius + middleH * 0.5, middleH * 0.3, M_PI_4, 3 * M_PI_4, 0);
    CGContextStrokePath(context);
    
    // 头发
    CGContextSetLineWidth(context, 1);
    // 中间
    CGContextMoveToPoint(context, rectW * 0.5, marginTop + 20);
    CGContextAddLineToPoint(context, rectW * 0.5, marginTop - 20);
    // 左边
    CGContextMoveToPoint(context, rectW * 0.5 -  10, marginTop + 20);
    CGContextAddLineToPoint(context, rectW * 0.5 - 30, marginTop - 20);
    // 最左边
    CGContextMoveToPoint(context, rectW * 0.5 - 20, marginTop + 20);
    CGContextAddLineToPoint(context, rectW * 0.5 - 50, marginTop - 20);
    // 右边
    CGContextMoveToPoint(context, rectW * 0.5 + 10, marginTop + 20);
    CGContextAddLineToPoint(context, rectW * 0.5 + 30, marginTop - 20);
    // 最右边
    CGContextMoveToPoint(context, rectW * 0.5 + 20, marginTop + 20);
    CGContextAddLineToPoint(context, rectW * 0.5 + 50, marginTop - 20);
    CGContextStrokePath(context);
}

@end

File: swift/Quartz2DDemo/Quartz2DDemo/HuangView.swift
//
//  HuangView.swift
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-2.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class HuangView: UIView {
    
    lazy var context = UIGraphicsGetCurrentContext()
    
    override func draw(_ rect: CGRect) {
        let marginTop:CGFloat = 50
        let radius:CGFloat = 100
        let middleH:CGFloat = 200
        
        // 设置黄色,绘制顶部，身体，底部
        context?.setFillColor(UIColor.yellow.cgColor)
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5, y: marginTop + radius), radius: CGFloat(radius), startAngle: 0, endAngle: CGFloat(Float.pi), clockwise: true)
        context?.addRect(CGRect.init(x: rect.width * 0.5 - radius, y: marginTop + radius, width: radius * 2, height: middleH))
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5, y: marginTop + radius + middleH), radius: radius, startAngle: 0, endAngle: CGFloat(Float.pi), clockwise: false)
        context?.fillPath()
        
        // 设置黑色，绘制眼眶
        context?.setFillColor(UIColor.black.cgColor)
        let padding:CGFloat = 5
        let glassH:CGFloat = 30
        context?.addRect(CGRect.init(x: rect.width * 0.5 - radius - padding, y: marginTop + radius, width: (radius + padding) * 2, height: glassH))
        context?.fillPath()
        
        // 设置白色，绘制大眼睛
        context?.setFillColor(UIColor.white.cgColor)
        let whiteRadius = radius * 0.3
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 - whiteRadius, y: marginTop + radius + glassH * 0.5), radius: whiteRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 + whiteRadius, y: marginTop + radius + glassH * 0.5), radius: whiteRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.fillPath()
        
        // 设置黑色，绘制黑眼圈
        context?.setStrokeColor(UIColor.black.cgColor)
        context?.setLineWidth(10)
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 - whiteRadius, y: marginTop + radius + glassH * 0.5), radius: whiteRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.move(to: CGPoint.init(x: rect.width * 0.5 + whiteRadius * 2, y: marginTop + radius + glassH * 0.5))
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 + whiteRadius, y: marginTop + radius + glassH * 0.5), radius: whiteRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.strokePath()
        
        // 设置棕色，绘制眼球
        context?.setFillColor(UIColor.brown.cgColor)
        let eyeballRadius = radius * 0.15
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 - whiteRadius, y: marginTop + radius + glassH * 0.5), radius: eyeballRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.move(to: CGPoint.init(x: rect.width * 0.5 + eyeballRadius * 2, y: marginTop + radius + glassH * 0.5))
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 + whiteRadius, y: marginTop + radius + glassH * 0.5), radius: eyeballRadius, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.fillPath()
        
        // 设置白色，绘制眼珠
        context?.setFillColor(UIColor.white.cgColor)
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 - whiteRadius, y: marginTop + radius + glassH * 0.5 - 5), radius: eyeballRadius * 0.2, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.move(to: CGPoint.init(x: rect.width * 0.5 + whiteRadius, y: marginTop + radius + glassH * 0.5 - 5))
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5 + whiteRadius, y: marginTop + radius + glassH * 0.5 - 5), radius: eyeballRadius * 0.2, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: false)
        context?.fillPath()
        
        // 设置黑色，绘制嘴巴
        context?.setStrokeColor(UIColor.black.cgColor)
        context?.setLineWidth(2)
        context?.addArc(center: CGPoint.init(x: rect.width * 0.5, y: marginTop + radius * 2), radius: 50, startAngle: CGFloat(Float.pi/4), endAngle: CGFloat(Float.pi/4 * 3), clockwise: false)
        context?.strokePath()
    }
}
